python/graphstorm/sagemaker/sagemaker_partition.py

```diff
@@ -0,0 +1,371 @@
+        for i in range(1, world_size):
+            client, _ = sock.accept()
+            client_list[i] = client
+    else:
+        # sync with master
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        for _ in range(30):
+            try:
+                sock.connect((leader_addr, 12000))
+                break
+            except: # pylint: disable=bare-except
+                logging.debug("Rank %s trying to connect to leader %s",
+                    host_rank, leader_addr)
+                time.sleep(10)
+        logging.debug("Connected")
+        client_list = None
+
+    # write ip list info to disk
+    ip_list = []
+    ip_list_path = os.path.join(tmp_data_path, 'ip_list.txt')
+    with open(ip_list_path, 'w', encoding='utf-8') as f:
+        for host in hosts:
+            f.write(f"{socket.gethostbyname(host)}\n")
+            ip_list.append(socket.gethostbyname(host))
+
+    graph_data_s3_no_trailing = graph_data_s3[:-1] if graph_data_s3.endswith('/') else graph_data_s3
+    graph_config_s3_path = f"{graph_data_s3_no_trailing}/{metadata_filename}"
+    meta_info_file = os.path.join(tmp_data_path, metadata_filename)
+
+    if not os.path.exists(meta_info_file):
+        if host_rank == 0:
+            logging.debug("Downloading metadata file from %s into %s",
+                graph_config_s3_path, meta_info_file)
+        download_data_from_s3(graph_config_s3_path, tmp_data_path,
+            sagemaker_session=sagemaker_session)
+
+    with open(meta_info_file, 'r') as f: # pylint: disable=unspecified-encoding
+        graph_config = json.load(f)
+
+    logging.info("Downloading graph data from %s into %s",
+        graph_data_s3, tmp_data_path)
+    graph_data_path = download_graph(
+        graph_data_s3,
+        graph_config,
+        world_size,
+        host_rank,
+        tmp_data_path,
+        sagemaker_session)
+
+    partition_config = PartitionerConfig(
+        metadata_file=meta_info_file,
+        local_output_path=tmp_data_path,
+        rank=host_rank,
+        sagemaker_session=sagemaker_session)
+
+    if job_config.partition_algorithm == 'random':
+        partitioner = RandomPartitioner(partition_config)
+    else:
+        raise RuntimeError(f"Unknown partition algorithm: '{job_config.partition_algorithm}'")
+
+    # Conditionally skip partitioning
+    if skip_partitioning:
+        # All workers + leader need to download partition data from S3
+        s3_partition_path = os.path.join(output_s3, "partition")
+        local_partition_path = os.path.join(graph_data_path, "partition")
+        logging.warning("Skipped partitioning step, trying to download partition data from %s",
+            s3_partition_path)
+
+        download_data_from_s3(s3_partition_path, local_partition_path, sagemaker_session)
+    else:
+        # Perform the partitioning. Leader uploads partition assignments to S3.
+        local_partition_path, s3_partition_path = partitioner.create_partitions(
```

Comment:
So now only leader do the random partition?

Reply:
For random partitioning yes we don't need to involve the workers as the partitioning takes seconds even for very large graphs. If we run into any performance issues we can revisit.

To be clear: `sagemaker_partition.py` is agnostic of the partitioning algorithm implementation, as long as it follows the contracts of the `PartitionAlgorithm` base class. 

One of the contracts is that at the end of the partition assignment step, the partition assignment files exist on the leader, which proceeds to upload them to the designated location.